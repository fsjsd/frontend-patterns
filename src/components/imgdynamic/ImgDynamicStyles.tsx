import styled from "styled-components";


export const ImageList = styled.div`
  display:inline-flex;
  flex-wrap: wrap;
  gap: 12px;
`;

export const ImageWrapper = styled.div<{ width?: string, height?: string }>`
  border:solid 0px transparent;
  border-radius:8px;
  clip-path: inset(0 0 0 0 round 8px);
  overflow:hidden;
  background-color:#ddd;
  width:${({ width }) => width ?? "100%"};
  height:${({ height }) => height ?? "100%"};
  &> img {
    border-radius:inherit;
    position:absolute;
  }
`;

export const ImageLoading = styled.div<{ width?: string, height?: string }>`
  width:${({ width }) => width ?? "100%"};
  height:${({ height }) => height ?? "100%"};
  background-color:#ddd;
  font-size:3rem;
  color:#d3d3d3;
  display:flex;
  align-items:center;
  justify-content:center;
`;
export const ImageError = styled.div<{ width?: string, height?: string }>`
  width:${({ width }) => width ?? "100%"};
  height:${({ height }) => height ?? "100%"};
  background-color:#fdd;
  font-size:3rem;
  color:#d78b8b;
  display:flex;
  align-items:center;
  justify-content:center;
`;

export default {};