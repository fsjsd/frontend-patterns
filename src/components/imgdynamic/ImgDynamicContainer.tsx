import React from 'react'
import { MdHourglassTop, MdOutlineError } from 'react-icons/md';
import { ContentContainer } from '../../ux/ContentContainer'
import ImgDynamic from './ImgDynamic'
import { ImageError, ImageList, ImageLoading, ImageWrapper } from './ImgDynamicStyles';

const imageSize = 200;
const imageSizePx = `${imageSize}px`;

// eslint-disable-next-line react/display-name
const Loading = React.memo(() => <ImageLoading width={imageSizePx} height={imageSizePx} >
  <MdHourglassTop />
</ImageLoading>)

// eslint-disable-next-line react/display-name
const Error = React.memo(() => <ImageError width={imageSizePx} height={imageSizePx} >
  <MdOutlineError />
</ImageError>)

/**
 * Container for list of ImgDynamic components
 * @returns {React.ReactElement}
 */
const ImgDynamicContainer = ({ imageCount }: { imageCount: number }) => {

  const testData = new Array(imageCount).fill(0);

  // Note this deliberately loads large images from picsum
  // to force longer loading times

  return (<>
    <ContentContainer
      title="Dynamic Images"
      codeLink="/features/imgdynamic"
      markDownPromise={import('./requirements.md')}
    >
      <div role="main">
        <ImageList>
          {testData.map(
            (_, i) => <ImageWrapper width={imageSizePx} height={imageSizePx} key={i}>
              <ImgDynamic
                id={`img-${i}`}
                loadingComp={<Loading />}
                errorComp={<Error />}
                width={imageSize}
                height={imageSize}
                src={`https://picsum.photos/id/${(Math.random() * 1000).toFixed(0)}/800`}
              />
            </ImageWrapper>
          )}
        </ImageList>
      </div>
    </ContentContainer>
  </>
  )
}

export default ImgDynamicContainer