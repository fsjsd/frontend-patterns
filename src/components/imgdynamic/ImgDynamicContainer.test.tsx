import React from 'react'
import ImgDynamicContainer from "./ImgDynamicContainer";
import { render, waitFor } from '@testing-library/react'
import { ImageError, ImageLoading, ImageWrapper } from './ImgDynamicStyles';

describe("ImgDynamic", () => {
  it("renders and matches snapshot", async () => {
    const { queryByRole, getByRole } = render(<ImgDynamicContainer imageCount={3} />)
    await waitFor(() => expect(queryByRole("main")).toBeInTheDocument());
    expect(getByRole("main")).toMatchSnapshot();
  })
  it("renders loading image without dimenstions", async () => {
    const { container } = render(<ImageLoading />)
    expect(container).toMatchSnapshot();
  })
  it("renders error image without dimenstions", async () => {
    const { container } = render(<ImageError />)
    expect(container).toMatchSnapshot();
  })
  it("renders error image without dimenstions", async () => {
    const { container } = render(<ImageWrapper />)
    expect(container).toMatchSnapshot();
  })

})