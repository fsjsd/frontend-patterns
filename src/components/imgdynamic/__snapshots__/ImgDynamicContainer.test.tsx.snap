// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImgDynamic renders and matches snapshot 1`] = `
.c2 {
  position: relative;
}

.c2> * {
  top: 0;
  left: 0;
  position: absolute;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 12px;
}

.c1 {
  border: solid 0px transparent;
  border-radius: 8px;
  -webkit-clip-path: inset(0 0 0 0 round 8px);
  clip-path: inset(0 0 0 0 round 8px);
  overflow: hidden;
  background-color: #ddd;
  width: 200px;
  height: 200px;
}

.c1> img {
  border-radius: inherit;
  position: absolute;
}

.c3 {
  width: 200px;
  height: 200px;
  background-color: #ddd;
  font-size: 3rem;
  color: #d3d3d3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (min-width:768px) {

}

<div
  role="main"
>
  <div
    class="c0"
  >
    <div
      class="c1"
      height="200px"
      width="200px"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
          height="200px"
          width="200px"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M6 2l.01 6L10 12l-3.99 4.01L6 22h12v-6l-4-4 4-3.99V2H6zm10 14.5V20H8v-3.5l4-4 4 4z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="c1"
      height="200px"
      width="200px"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
          height="200px"
          width="200px"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M6 2l.01 6L10 12l-3.99 4.01L6 22h12v-6l-4-4 4-3.99V2H6zm10 14.5V20H8v-3.5l4-4 4 4z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="c1"
      height="200px"
      width="200px"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
          height="200px"
          width="200px"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M6 2l.01 6L10 12l-3.99 4.01L6 22h12v-6l-4-4 4-3.99V2H6zm10 14.5V20H8v-3.5l4-4 4 4z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ImgDynamic renders error image without dimenstions 1`] = `
.c0 {
  width: 100%;
  height: 100%;
  background-color: #fdd;
  font-size: 3rem;
  color: #d78b8b;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`ImgDynamic renders error image without dimenstions 2`] = `
.c0 {
  border: solid 0px transparent;
  border-radius: 8px;
  -webkit-clip-path: inset(0 0 0 0 round 8px);
  clip-path: inset(0 0 0 0 round 8px);
  overflow: hidden;
  background-color: #ddd;
  width: 100%;
  height: 100%;
}

.c0> img {
  border-radius: inherit;
  position: absolute;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`ImgDynamic renders loading image without dimenstions 1`] = `
.c0 {
  width: 100%;
  height: 100%;
  background-color: #ddd;
  font-size: 3rem;
  color: #d3d3d3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  />
</div>
`;
